sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java

```diff
@@ -363,4 +317,135 @@ private R visitQueriedRelation(QueriedRelation relation, Field field) {
             return symbolVisitor.process(output, null);
         }
     }
+
+    private static class SubselectRewriter extends AnalyzedRelationVisitor<RelationNormalizer.Context, AnalyzedRelation> {
+
+        private static final SubselectRewriter SUBSELECT_REWRITER = new SubselectRewriter();
+
+        private SubselectRewriter() {
+        }
+
+        public static AnalyzedRelation rewrite(AnalyzedRelation relation, Context context) {
+            return SUBSELECT_REWRITER.process(relation, context);
+        }
+
+        @Override
+        protected AnalyzedRelation visitAnalyzedRelation(AnalyzedRelation relation, Context context) {
+            return relation;
+        }
+
+        @Override
+        public AnalyzedRelation visitQueriedSelectRelation(QueriedSelectRelation relation, Context context) {
+            QuerySpec querySpec = relation.querySpec();
+            // Try to merge with parent query spec
+            if (canBeMerged(querySpec, context.querySpec)) {
+                querySpec = mergeQuerySpec(querySpec, context.querySpec);
+            }
+
+            // Try to push down to the child
+            context.querySpec = querySpec;
+            AnalyzedRelation processedChildRelation = process(relation.relation(), context);
+
+            // If cannot be pushed down replace qSpec with possibly merged qSpec from context
+            if (processedChildRelation == null) {
+                relation.querySpec(context.querySpec);
```

Comment:
so yep the process() results in a mutation because of replaceFieldReferences() which should be down only by "concrete" relations...